lib/auth/keystore/gcp_kms.go

```diff
@@ -0,0 +1,454 @@
+		if err != nil && !strings.Contains(err.Error(), "has value DESTROY") {
+			g.log.WithFields(logrus.Fields{"key_version": unusedKey.keyVersionName}).WithError(err).Warn("error deleting unused GCP KMS key")
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// kmsSigner implements the crypto.Signer interface
+type kmsSigner struct {
+	kmsClient kmsClientI
+	keyID     gcpKMSKeyID
+	public    crypto.PublicKey
+	clock     clockwork.Clock
+}
+
+func (g *gcpKMSKeyStore) newKmsSigner(ctx context.Context, keyID gcpKMSKeyID) (*kmsSigner, error) {
+	req := &kmspb.GetPublicKeyRequest{
+		Name: keyID.keyVersionName,
+	}
+	// Retry fetching the public key while the key is pending creation.
+	resp, err := retryWhilePending(ctx, g.clock, g.kmsClient.GetPublicKey, req)
+	if err != nil {
+		return nil, trace.Wrap(err, "unexpected error fetching public key")
+	}
+
+	block, _ := pem.Decode([]byte(resp.Pem))
+	if block == nil {
+		return nil, trace.BadParameter("GCP KMS key %s has invalid public key PEM", keyID.keyVersionName)
+	}
+	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return nil, trace.Wrap(err, "unexpected error parsing public key pem")
+	}
+
+	return &kmsSigner{
+		kmsClient: g.kmsClient,
+		clock:     g.clock,
+		keyID:     keyID,
+		public:    pub,
+	}, nil
+}
+
+func (s *kmsSigner) Public() crypto.PublicKey {
+	return s.public
+}
+
+func (s *kmsSigner) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	ctx := context.Background()
+
+	requestDigest := &kmspb.Digest{}
+	switch opts.HashFunc() {
+	case crypto.SHA256:
+		requestDigest.Digest = &kmspb.Digest_Sha256{
+			Sha256: digest,
+		}
+	case crypto.SHA512:
+		requestDigest.Digest = &kmspb.Digest_Sha512{
+			Sha512: digest,
+		}
+	default:
+		return nil, trace.BadParameter("unsupported hash func for GCP KMS signer: %v", opts.HashFunc())
+	}
+
+	resp, err := doGCPRequest(ctx, s.clock, s.kmsClient.AsymmetricSign, &kmspb.AsymmetricSignRequest{
+		Name:   s.keyID.keyVersionName,
+		Digest: requestDigest,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "error while attempting GCP KMS signing operation")
+	}
+	return resp.Signature, nil
+}
+
+type gcpKMSKeyID struct {
+	keyVersionName string
+}
+
+func (g gcpKMSKeyID) marshal() []byte {
+	var buf bytes.Buffer
+	buf.WriteString(gcpkmsPrefix)
+	buf.WriteString(g.keyVersionName)
+	return buf.Bytes()
+}
+
+func parseGCPKMSKeyID(key []byte) (gcpKMSKeyID, error) {
+	var keyID gcpKMSKeyID
+	if keyType(key) != types.PrivateKeyType_GCP_KMS {
+		return keyID, trace.BadParameter("unable to parse invalid GCP KMS key")
+	}
+	// strip gcpkms: prefix
+	keyID.keyVersionName = string(key[len(gcpkmsPrefix):])
+	return keyID, nil
+}
+
+// withTimeout is like context.WithTimeout, but with mockable time
+func withTimeout(ctx context.Context, clock clockwork.Clock, d time.Duration) (context.Context, func()) {
+	ctx, ctxCancel := context.WithCancel(ctx)
+	timer := clock.NewTimer(d)
```

Comment:
Use `clock.After(d)` instead of a timer since you only need it to tick at most once.

Same goes for below.

Reply:
Is there a way to distinguish a real `clockwork` clock vs a fake one? `context.WithTimeout` will actually return a context for which `.Deadline()` is nonzero (often used by networking libraries to set OS-level deadlines and timeouts), and doesn't spawn extra goroutines if they can be avoided.

Reply:
This is a timer, not a ticker. It seems a bit better, the docs on `time.After`:
>  If efficiency is a concern, use NewTimer instead and call Timer.Stop if the timer is no longer needed.

It is certainly possible to detect a real clock and just use `context.WithTimeout` in that case, I'll go ahead and do that

Reply:
Thanks for pointing that out Edoardo, I made it handle `Deadline()` calls correctly as well. FWIW `context.WithTimeout` does spawn at least 1 goroutine, but we can save on the extra 2 from `context.WithCancel` and the timer by calling it directly, so I added a "real" clock distinction to the api and optimized for that case